### ISSUE TITLE
Error occurred when executing LoRA Caption Save:  'list' object has no attribute 'strip'

### ISSUE THREAD

**Issue by Nebuluss** [LarryJane491/Image-Captioning-in-ComfyUI#13]
[)](
![Example](https://github.com/user-attachments/assets/81753151-feda-4a13-828e-31feb79d5514)
)

![image](https://github.com/user-attachments/assets/873cde53-fcfe-4e42-9c40-f1e9aa1a3f1f)

Error occurred when executing LoRA Caption Save:

'list' object has no attribute 'strip'

File "D:\ANEWCOMFY\ComfyUI_windows_portable\ComfyUI\execution.py", line 152, in recursive_execute
output_data, output_ui = get_output_data(obj, input_data_all)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "D:\ANEWCOMFY\ComfyUI_windows_portable\ComfyUI\execution.py", line 82, in get_output_data
return_values = map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "D:\ANEWCOMFY\ComfyUI_windows_portable\ComfyUI\execution.py", line 75, in map_node_over_list
results.append(getattr(obj, func)(**slice_dict(input_data_all, i)))
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "D:\ANEWCOMFY\ComfyUI_windows_portable\ComfyUI\custom_nodes\Image-Captioning-in-ComfyUI\LoRAcaption.py", line 40, in save_text_file
if text.strip() == '':
^^^^^^^^^^

**Comment by Nebuluss** [comment]
Caption Splitting and File Saving:

Issue: All captions were previously being saved into a single text file.
Change: The code was updated to split the text input into individual captions based on line breaks (\n). Each caption is now saved into a separate text file, corresponding to its respective image.
Filename Matching:

Issue: The captions were not being properly associated with their respective image filenames.
Change: The namelist (list of image filenames) is now split into individual filenames, and each caption is matched with the correct filename. Each caption is saved with the corresponding image name as a .txt file.
Prefix Handling:

Issue: The prefix handling logic needed to ensure that captions were formatted correctly.
Change: The prefix is now correctly appended to each caption before it is saved, ensuring consistent formatting in the output files.
Error Handling for Mismatched Captions and Filenames:

Issue: There was no handling for cases where the number of captions didn’t match the number of filenames.
Change: A check was added to ensure that the number of captions matches the number of filenames. If they don't match, the process is halted with an error message to prevent incorrect file generation.
Refactoring for Better Readability and Maintenance:

Change: The code was refactored to improve readability and make it easier to maintain. This included splitting functions and adding comments to clarify each part of the process.

**Comment by Nebuluss** [comment]
import glob
import os
from PIL import Image, ImageOps
import numpy as np
import torch
import comfy

class LoRACaptionSave:
    def __init__(self):
        pass

    @classmethod
    def INPUT_TYPES(cls):
        return {
            "required": {
                "namelist": ("STRING", {"forceInput": True}),
                "path": ("STRING", {"forceInput": True}),
                "text": ("STRING", {"forceInput": True}),
            },
            "optional": {
                "prefix": ("STRING", {"default": " "}),
            }
        }

    OUTPUT_NODE = True
    RETURN_TYPES = ()
    FUNCTION = "save_text_file"
    CATEGORY = "LJRE/LORA"

    def save_text_file(self, text, path, namelist, prefix):
        # Ensure the path exists
        if not os.path.exists(path):
            print(f"The path `{path}` doesn't exist! Creating it...")
            try:
                os.makedirs(path, exist_ok=True)
            except OSError as e:
                print(f"The path `{path}` could not be created! Is there write access?\n{e}")
                return

        # Check if 'text' is a list and join it into a single string
        if isinstance(text, list):
            text = '\n'.join(text)
        
        # Split text into individual captions based on line breaks
        captions = text.split('\n')

        # Split the namelist into individual filenames
        namelistsplit = namelist.splitlines()
        namelistsplit = [i[:-4] for i in namelistsplit]

        if len(captions) != len(namelistsplit):
            print("The number of captions does not match the number of filenames.")
            return

        for i, caption in enumerate(captions):
            # Handle the prefix for each caption
            if prefix.endswith(","):
                caption = prefix + " " + caption
            elif not prefix.endswith(", "):
                caption = prefix + ", " + caption

            # Generate filename and save each caption to a separate text file
            filename = f"{namelistsplit[i]}.txt"
            file_path = os.path.join(path, filename)
            self.write_text_file(file_path, caption)

        return (text, {"ui": {"string": text}})

    def write_text_file(self, file, content):
        try:
            with open(file, 'w', encoding='utf-8', newline='\n') as f:
                f.write(content)
        except OSError:
            print(f"Unable to save file `{file}`")

def io_file_list(dir='', pattern='*.txt'):
    return glob.glob(os.path.join(dir, pattern))

class LoRACaptionLoad:
    def __init__(self):
        pass

    @classmethod
    def INPUT_TYPES(cls):
        return {
            "required": {
                "path": ("STRING", {"default": ""}),
            },
        }

    RETURN_TYPES = ("STRING", "STRING", "IMAGE",)
    RETURN_NAMES = ("Name list", "path", "Image list",)

    FUNCTION = "captionload"
    CATEGORY = "LJRE/LORA"

    def captionload(self, path, pattern='*.png'):
        text = io_file_list(path, pattern)
        text = list(map(os.path.basename, text))
        text = '\n'.join(text)

        if not os.path.isdir(path):
            raise FileNotFoundError(f"Path '{path}' cannot be found.")
        dir_files = os.listdir(path)
        if not dir_files:
            raise FileNotFoundError(f"No files in path '{path}'.")

        # Filter files by extension
        valid_extensions = ['.png']
        dir_files = [f for f in dir_files if any(f.lower().endswith(ext) for ext in valid_extensions)]

        dir_files = [os.path.join(path, x) for x in dir_files]

        images = []
        image_count = 0

        for image_path in dir_files:
            if os.path.isdir(image_path):
                continue
            i = Image.open(image_path)
            i = ImageOps.exif_transpose(i)
            image = i.convert("RGB")
            image = np.array(image).astype(np.float32) / 255.0
            image = torch.from_numpy(image)[None,]
            images.append(image)
            image_count += 1

        if len(images) == 1:
            return (images[0], 1)
        elif len(images) > 1:
            image1 = images[0]
            for image2 in images[1:]:
                if image1.shape[1:] != image2.shape[1:]:
                    image2 = comfy.utils.common_upscale(image2.movedim(-1, 1), image1.shape[2], image1.shape[1], "bilinear", "center").movedim(1, -1)
                image1 = torch.cat((image1, image2), dim=0)

        return text, path, image1, len(images)
